Sources/SwiftLexicalLookup/Scopes/ScopeImplementations.swift

```diff
@@ -503,17 +415,46 @@ import SwiftSyntax
           checkIdentifier(identifier, refersTo: name, at: lookUpPosition)
         }
 
-      return (implicitSelf.isEmpty ? [] : [.fromScope(self, withNames: implicitSelf)]) + [.lookInMembers(self)]
-        + defaultLookupImplementation(identifier, at: lookUpPosition, with: config)
+      return LookupResult.getResultArray(for: self, withNames: implicitSelf)
+        + [.lookInGenericParametersOfExtendedType(self)]
+        + defaultLookupImplementation(identifier, at: lookUpPosition, with: config, propagateToParent: false)
+        + [.lookInMembers(self)]
+        + lookupInParent(identifier, at: lookUpPosition, with: config)
```

Comment:
That's a lot of array concatenation, but I like how it very clearly shows each of the necessary lookup steps in the appropriate order. (No action needed)

Reply:
I also like how it shows the steps, but I think we could try to improve readability here a bit. I think I’d like to try to rewrite it to some kind of result builder method at some point in the future when we have a good suite of integration and unit tests (I think I’ve mentioned it once during a meeting). I was thinking about something like this:

```swift
return lookupResult(identifier, at: lookupPosition, with: config)
  .implicitSelf()
  .lookInGenericParametersOfExtendedType()
  .defaultLookup()
  .lookInMembers()
  .lookupInParent()
```

Each of the methods could also take an optional `if: bool` parameter that would conditionally include that step. This way we could basically remove all those complicated `if … else if … else` statements from `lookup` methods. Additionally, we could try to enforce some pre/post conditions related to `config` in those functions (like `LookupConfig.finishInSequentialScope` being checked inside `.lookupInParent()` call for all sequential scopes).

Reply:
That does look nice! I don't know whether it's "worth" the effort of doing (vs. other things we've talked about for this library), but the result would be cleaner code in this part of the library.